### PR TITLE
Blog post on Spring Microservices with OAuth

### PR DESCRIPTION
**Pull request by mraible**
Uses Spring Security and Spring Cloud Security, no Okta SDKs.

### CODE REVIEW

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+
+```properties
+zuul.routes.beer-catalog-service.path=/beers
+zuul.routes.beer-catalog-service.url=http://localhost:8080
+
+zuul.routes.home.path=/home
+zuul.routes.home.url=http://localhost:8080
+```
+
+Open `edge-service/src/main/java/com/example/EdgeServiceApplication.java` and add `@EnableOAuth2Sso` to enable single sign-on (SSO) with OAuth.
```
Comment (nbarbettini) [inline _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: Would it be valid to say "enable authentication with OAuth" instead? This isn't really SSO in the sense that Okta normally refers to SSO. (If you'd rather stick with the Spring Security nomenclature, that's fine)

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+
+Restart your Edge Server application, navigate to `http://localhost:8081/home` and you'll see your user details on the next page.
+
+{% img blog/microservices-spring-oauth/user-details.png alt:"Okta User Details" width:"800" %}{: .center-image }
+
+**NOTE:** I was unable to get the logout button to work due to a CSRF error. I tried adding `security.enable-csrf=false` to `application.properties` in the Edge Service app, but it didn't help. I sent an email to the Spring Security team asking if they had any advice.
+
+## Add Okta's Sign-In Widget to the Angular Client
+
+To use Okta's Sign-In Widget, you'll need to modify your app in Okta to enable an *Implicit* grant type. Log in to your account, navigate to **Applications** > **Spring OAuth** > **General** tab and click **Edit**. Enable **Implicit (Hybrid)** under **Allowed grant types** and select both checkboxes below it. Add `http://localhost:4200` under **Login redirect URIs** and click **Save**.
```
Comment (nbarbettini) [inline _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: **the** Implicit grant type

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+import * as OktaSignIn from '@okta/okta-signin-widget';
+
+@Injectable()
+export class OktaService {
+  widget;
+
+  constructor() {
+    this.widget = new OktaSignIn({
+      baseUrl: 'https://{yourOktaDomain}.com',
+      clientId: '{yourClientId}',
```
Comment (nbarbettini) [inline _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: `{clientId}`

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+        config.setAllowedMethods(Collections.singletonList("*"));
+        config.setAllowedHeaders(Collections.singletonList("*"));
+        source.registerCorsConfiguration("/**", config);
+        FilterRegistrationBean bean = new FilterRegistrationBean(new CorsFilter(source));
+        bean.setOrder(Ordered.HIGHEST_PRECEDENCE);
+        return bean;
+    }
+}
+```
+
```
Comment (nbarbettini) [inline _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: Think we should mention that you should tone back `Allow: *` once you're ready to go to production?

File: _source/_posts/2017-06-15-build-microservices-architecture-spring-boot.md [hunk 5]
```diff
@@ -476,7 +476,8 @@ Content-Type: application/json;charset=UTF-8
 Date: Wed, 07 Feb 2018 16:51:38 GMT
 Transfer-Encoding: chunked
 X-Application-Context: edge-service:8081
-
+```
```
Comment (nbarbettini) [inline _source/_posts/2017-06-15-build-microservices-architecture-spring-boot.md]: Stray grave?

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+    public void configure(HttpSecurity http) throws Exception {
+        http
+            .requestMatcher(new RequestHeaderRequestMatcher("Authorization"))
+            .authorizeRequests()
+            .antMatchers("/**").authenticated();
+    }
+}
+```
+
+At this point, you've configured enough to sign-in to your Edge Service application, but it won't be able to communicate with the downstream beer-catalog-service.
```
[the remark this answers is not in the document]
Comment (mraible) [reply _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: Will fix.

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md (hunk 1 above)
Comment (mraible) [reply _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: Good point.

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md (hunk 4 above)
Comment (mraible) [reply _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: Will add:

**NOTE:** You should change the allowed origins from `*` to your client's URLs if you're using this configuration in production.

File: _source/_posts/2017-06-15-build-microservices-architecture-spring-boot.md (hunk 5 above)
Comment (mraible) [reply _source/_posts/2017-06-15-build-microservices-architecture-spring-boot.md]: No, this is correct. I like to split up the command and the response into separate code blocks. The full code in this is as follows:

```bash
$ http localhost:8081/good-beers
HTTP/1.1 200
Content-Type: application/json;charset=UTF-8
Date: Wed, 07 Feb 2018 16:50:51 GMT
Transfer-Encoding: chunked
X-Application-Context: edge-service:8081
```
```json
[
    {
        "name": "Kentucky Brunch Brand Stout"
    },
    {
        "name": "Good Morning"
    },
    {
        "name": "Very Hazy"
    },
    {
        "name": "King Julius"
    }
]
```

Comment (nbarbettini) [reply _source/_posts/2017-06-15-build-microservices-architecture-spring-boot.md]: Ah, gotcha. Nevermind!

File: _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md (hunk 3 above)
Comment (nbarbettini) [reply _source/_posts/2018-02-13-secure-spring-microservices-with-oauth.md]: AFAIK, no. I can't remember if we ever landed on a consistent style to use. `{yourClientId}` is fine 😄

**Pull request merged**